docs/release-notes/mongoid-8.0.txt

```diff
@@ -57,6 +57,45 @@ changed in Mongoid 8.0:
 Please refer to :ref:`configuration option <configuration-options>` for
 the description and effects of each of these options.
 
+Storing Uncastable Value
+------------------------
+
+In Mongoid 8, Mongoid standardizes the storing of "uncastable values." On
+attempting to write an uncastable value, a ``nil`` is written instead. See the
+secion on :ref:`Uncastable Values <uncastable-values>` for more details.
+
+Some ``mongoize`` methods were also changed to perform consistently with rails
+and the other mongoize methods. The following is a table of the changes in
+functionality:
+
++--------------+------------------------+------------------------+-------------------+
+| Field Type   | Situation              | Previous Functionality | New Functionality |
++==============+========================+========================+===================+
+| Boolean      | When a non-boolean     | return ``false``       | return ``nil``    |
+|              | string is assigned:    |                        |                   |
+|              | "bogus value"          |                        |                   |
++--------------+------------------------+------------------------+-------------------+
+| Symbol       | When a value that does | return ``nil``         | return ``nil``    |
+|              | not respond to         |                        |                   |
+|              | ``to_sym`` is          |                        |                   |
+|              | assigned: ``[]``       |                        |                   |
++--------------+------------------------+------------------------+-------------------+
+| Array/Set    | When a value that is   | return ``[ 1, 2, 3 ]`` | return ``nil``    |
+|              | not an array or set is | return ``[[ :x, 1 ]]`` | return ``nil``    |
+|              | assigned, but does     |                        |                   |
+|              | respond to ``to_a``:   |                        |                   |
+|              | ``1..3``, ``{ x: 1 }`` |                        |                   |
```

Comment:
I would add in the second column something like

> While it doesn't usually make sense to convert Range and Hash instances to Arrays or Sets, instances of custom classes implementing to_a that were assigned to Array/Set would now be discarded. Please explicitly call ``to_a`` on the custom classes when assigning to Array/Set fields.

And we might need to go back on this change if we receive user feedback that it's problematic, this might be something to highlight in release notes for RC which would make sense to do.

Reply:
If we have to tell users to do this *or else they will silently lose data*, it's a clear indication that the change is problematic. At a minimum it needs a *feature flag*, but better not to do it at all IMHO.

Reply:
Excuse me, this documentation is incorrect. Our current functionality raises on any non-array value to an array field:
```
irb(main):001:1* class User
irb(main):002:1*   include Mongoid::Document
irb(main):003:1*   field :arr, type: Array
irb(main):004:0> end
=> #<Mongoid::Fields::Standard:0x00007f9138758b10 @name="arr", @options={:type=>Array, :klass=>User}, @label=nil, @default_val=nil>
irb(main):005:0> user = User.new
=> #<User _id: 62ca6c1f3282a42670df183c, arr: nil>
irb(main):006:0> user.arr = 1..3
Mongoid::Errors::InvalidValue ()
message:
  Value of type Range cannot be written to a field of type Array
summary:
  Tried to set a value of type Range to a field of type Array
resolution:
  Verify if the value to be set correspond to field definition
```
I would be okay leaving this in if we want don't want to change that but otherwise I think we're clear on the Range/Hash to_a situation.

Reply:
@Neilshweky well that changes things! In legacy code, it seems the `InvalidValue` error is *only* used for Array/Hash.

After some further consideration, **I'd propose to add back `InvalidValue` error behavior**, pending further thinking about `strict` etc. handling.

My reasoning is that, since I expect Range/Hash to be casted (they define `.to_a`), it's better to raise an error for now than silently drop the data. In particular I'm worried about users doing mass-assignment in Rails, where they may not be aware of the types which ActionController is giving them.

AFAIK none of the other changes in this PR increase the chance for "data loss" vs. the legacy state. For example, casting Boolean "Foobar" to `nil` (new) rather than `false` (legacy) is not anymore loss-y than the previous state.

Reply:
Im not completely opposed to raising for array and hash for consistency. @p-mongo what do you think

Reply:
The current behavior in this PR (change from raising InvalidValue to discarding input and assigning nil) is OK with me. The raising on invalid input is covered by https://jira.mongodb.org/browse/MONGOID-5382.

@johnnyshields  I suggest you formulate your thoughts for raising on assignment to Array/Hash fields in a way that is aligned with the requirements/proposal in https://jira.mongodb.org/browse/MONGOID-5382 and add them there.

Reply:
Commented here: MONGOID-5222. The core reason *not* to do the change in this PR is that the logic in question is in `mongoid/attributes.rb` and is not part of any `mongoize` method, which makes it out-of-scope for the original ticket. 5222 specifically mentions `mongoize`.

(I believe the code being located outside `mongoize` is the reason both Neil and I were confused about the spec above.)

I'd like to think further about how exactly we handle MONGOID-5382 before commenting there, but we should defer any change to `mongoid/attributes.rb` until then.